File: catkin_ws/src/cone_detector/src/echo.cpp
#include <iostream>
#include <ros/ros.h>
#include <stdlib.h>
#include <stdio.h>
#include <image_transport/image_transport.h>
#include <sensor_msgs/image_encodings.h>
#include <cv_bridge/cv_bridge.h>
using namespace cv;
using namespace std;
Mat image;

void imageCallback(const sensor_msgs::ImageConstPtr& msg)
{
        image = cv_bridge::toCvCopy(msg, "bgr8")->image;
}

int main(int argc, char **argv)
{
    ros::init(argc, argv, "image_listener");
    ros::NodeHandle nh;
    image_transport::ImageTransport it(nh);
    image_transport::Subscriber sub = 
        it.subscribe("/image_raw", 1, imageCallback);
    image_transport::Publisher  pub = it.advertise("echo_image", 1 );
    sensor_msgs::ImagePtr msg ;
    while (nh.ok()) {
        if (!image.empty()){
            msg = cv_bridge::CvImage(
                    std_msgs::Header(), "bgr8", image).toImageMsg();
            pub.publish(msg);
            }
        ros::spinOnce();
        }
  return 0;


 }

  

File: catkin_ws/src/cone_detector/src/echo.py
#!/usr/bin/env python
import cv2
import rospy
from sensor_msgs.msg import Image
from cv_bridge import CvBridge, CvBridgeError
import threading


class Echo:
    def __init__(self, target_img="cone.png"):
        self.node_name = "Echo"
        self.thread_lock = threading.Lock()
        self.sub_image = rospy.Subscriber("~image_raw",\
                Image, self.cbImage, queue_size=1)
        self.pub_image = rospy.Publisher("~echo_image",\
                Image, queue_size=1)
        self.bridge = CvBridge()

        rospy.loginfo("[%s] Initialized." %(self.node_name))

    def cbImage(self,image_msg):
        thread = threading.Thread(target=self.processImage,args=(image_msg,))
        thread.setDaemon(True)
        thread.start()


    def processImage(self, image_msg):
        if not self.thread_lock.acquire(False):
            return
        image_cv = self.bridge.imgmsg_to_cv2(image_msg)
        
        try:
            self.pub_image.publish(\
                    self.bridge.cv2_to_imgmsg(image_cv, "bgr8"))
        except CvBridgeError as e:
            print(e)
        self.thread_lock.release()


if __name__=="__main__":
    rospy.init_node('Echo')
    e = Echo()
    rospy.spin()

